### DISCUSSION TITLE
Problem when running `start_egm_joint`

### DISCUSSION THREAD

**Discussion by adkoessler** [ros-industrial/abb_robot_driver#76] [General]
Hi,

I'm using RobotStudio 2022 and RobotWare 6.14 to control an IRB1200 robot (in a virtual station) using ROS. I've had good progress in the tutorials and RWS works fine. However, when I'm trying to execute [example 2](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_robot_bringup_examples#example-2-ex2_rws_and_egm_6axis_robot), I cannot move the simulated robot through ROS.

The UDPUC transmission protocols have been set up as show in tutorials. First of all, it is not clear to me what RAPID instructions should be fed within the T_ROB1 task. I tried using the code below, which was attached to some messages on this GitHub. 
```
MODULE TRob1Main
!======================================================================================================
! Copyright (c) 2018, ABB Schweiz AG
! All rights reserved.
!
! Redistribution and use in source and binary forms, with
! or without modification, are permitted provided that
! the following conditions are met:
!
!    * Redistributions of source code must retain the
!      above copyright notice, this list of conditions
!      and the following disclaimer.
!    * Redistributions in binary form must reproduce the
!      above copyright notice, this list of conditions
!      and the following disclaimer in the documentation
!      and/or other materials provided with the
!      distribution.
!    * Neither the name of ABB nor the names of its
!      contributors may be used to endorse or promote
!      products derived from this software without
!      specific prior wrSitten permission.
!
! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
! ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
! LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
! THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
!======================================================================================================
    
    !***********************************************************
    ! Program data
    !***********************************************************
    ! Home position.
    LOCAL CONST jointtarget home := [[0, 0, 0, 0, 30, 0], [9E9, 9E9, 9E9, 9E9, 9E9, 9E9]];
    
    ! Identifier for the EGM correction.
    LOCAL VAR egmident egm_id;
    
    ! Limits for convergance.
    LOCAL VAR egm_minmax egm_condition := [-0.1, 0.1];
            
    !***********************************************************
    !
    ! Procedure main
    !
    !   This RAPID code exemplify how to run EGM joint position
    !   motions.
    !
    !   Note: Update the UCDevice "ROB_1" with correct
    !         values for the remote address and port
    !         (i.e. to the EGM server).
    !
    !         Update via RobotStudio:
    !         Controller tab -> Configuration ->
    !         Communication -> Transmission Protocol
    !
    !***********************************************************
    PROC main()
        WHILE TRUE DO
            MoveAbsJ home, v200, fine, tool0;
            
            ! Register an EGM id.
            EGMGetId egm_id;
            
            ! Setup the EGM communication.
            EGMSetupUC ROB_1, egm_id, "default", "ROB_1", \Joint;
            
            ! Prepare for an EGM communication session.
            EGMActJoint egm_id
                        \J1:=egm_condition
                        \J2:=egm_condition
                        \J3:=egm_condition
                        \J4:=egm_condition
                        \J5:=egm_condition
                        \J6:=egm_condition
                        \MaxSpeedDeviation:=20.0;
                        
            ! Start the EGM communication session.
            EGMRunJoint egm_id, EGM_STOP_RAMP_DOWN, \J1 \J2 \J3 \J4 \J5 \J6 \CondTime:=5 \RampOutTime:=5;
            
            ! Release the EGM id.
            EGMReset egm_id;
            
            WaitTime 5;
        ENDWHILE
        
    ERROR
        IF ERRNO = ERR_UDPUC_COMM THEN
            TPWrite "Communication timedout";
            TRYNEXT;
        ENDIF
    ENDPROC
ENDMODULE
```
When starting the RAPID code, the status of EGM channel switches between active (true) and inactive (false). I don't know whether this is expected. Issuing `rosservice call /rws/sm_addin/start_egm_joint "{}"` then results in an error. I tried searching for this error, but had little success doing so.

Any input is greatly appreciated.

**Comment by gavanderhoorn** [comment]
The general idea is that *if* you're using `abb_robot_driver`, you'd be using the StateMachine addon.

If you don't want to/can't use the addon, you're responsible for writing the RAPID code. But at that point you can't call `/rws/sm_addin/...` services any more, as you're not using it.

(I'm confused why you would have the `sm_addin` set of functions available though, because IIRC, `abb_robot_driver` detects at runtime whether you have the addon or not. If not, you shouldn't see the service servers)

You appear to be using `abb_robot_driver`, so I'd expect you'd use the addon. Are you using the addon?

PS: this appears to be a question about `abb_robot_driver` setup, not about something specific to `abb_libegm`. Did you have a particular reason for posting on this tracker/Discussion board?

**Reply by adkoessler** [reply]
Thank you @gavanderhoorn for your answer. If you feel like this does not belong here, I can close this discussion and post it to `abb_robot_driver` instead, just let me know.

Yes I have the StateMachine 1.1 addon enabled. From what I understand from your message, one can either (i) use rapid code to manage the EGM channel and perform motions or (ii) use the SM addon, and then you can perform the motions while having no RAPID code running in your controller. Is it right?

I'm trying to perform the sequence of commands shown on [example 2](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_robot_bringup_examples#example-2-ex2_rws_and_egm_6axis_robot). Whether I add `TRob1Main.mod` to the task `T_ROB1` or not, the following error appears:
```
$ rosservice call /rws/sm_addin/start_egm_joint "{}"
ERROR: Service [/rws/sm_addin/start_egm_joint] is not available.
```
(Of course I cannot `start_rapid` when there is no RAPID to be executed.)

**Reply by gavanderhoorn** [reply]
> Yes I have the StateMachine 1.1 addon enabled

then I'm not sure I understand what you're trying to achieve with the RAPID code you showed.

The Addon should have generated all the necessary code for you.

**Comment by adkoessler** [comment]
Thank you @gavanderhoorn for providing me with an answer. I indeed missed steps regarding the usage of StateMachine add-in. Following the user manual of the add-in ("System Installation" section) to setup the controller solved the problem.

**Reply by gavanderhoorn** [reply]
Good to hear you got it to work.
